### ISSUE TITLE
The ability to run JS code periodically

### ISSUE THREAD

**Issue by xeioex** [nginx/njs#660]
Hi @hongzhidao, @drsm, @jirutka, @ivanitskiy

[Here](https://gist.github.com/xeioex/0d469fe881302816667f126458191853) is a new feature in progress that allows to run arbitrary JS code periodically without a client HTTP request or TCP connection. The feature allows to do some preparation work during startup (certs loading) or a keeper activities (keys rotation) during runtime.

Currently the patch is in alpha stage, nevertheless it has tests that cover timers and ngx.fetch(). Once the HTTP is ready I plan to add a stream version as well.

```
example.conf:
    location @periodics {
        # Specifies a JS handler to be run at 1 minute intervals
        js_periodic main.handler interval=60s jitter=5s;

        resolver 10.0.0.1;
        js_fetch_trusted_certificate /path/to/ISRG_Root_X1.pem;
    }

example.js:
    async function handler() {
        if (ngx.worker_id != 0) {
            /* using ngx.worker_id to run handler only in one worker. */
            return;
        }

        let reply = async ngx.fetch('https://nginx.org/en/docs/njs/');
        let body = async reply.text();

        ngx.log(ngx.INFO, body);
    }
```

Feel free to give your feedback and report bugs.

**Comment by hongzhidao** [comment]
Hi @xeioex 
Great job, this feature is very powerful, especially with the shared dictionary.
Do you think it's necessary to make it programmable to stop a periodic? For example:
```
function handler(periodic) {
        ...
        periodic.stop();  /* It will not add timer */
    }
```
If yes, we need to rethink the use, and it may change the implementation.
1. It will need to re-add a periodic unless there might be no periodic in the worker process.
2. Creating a new periodic can happen in any JS code, for example in the function related to js_content.

**Comment by xeioex** [comment]
Hi @hongzhidao,

> Do you think it's necessary to make it programmable to stop a periodic? 

Yes, agree this makes sense. Especially given that most of the time we need work in just one worker not in all, so we will be able just to disable all the periodics in other workers. I will add it.

>Creating a new periodic can happen in any JS code, for example in the function related to js_content.

I am not sure that we really need it. If a user is allowed to create an arbitrary number of timer events not linked to any request that may be a cause of DOS or CPU hog. The alternative solution can be a shared queue that requests are able to push their task messages to be processed and and a single periodic handler per worker that processes the events.

**Comment by hongzhidao** [comment]
> I will add it.

That will be great. Btw, does it mean you will add a prototype for it like `r` to a http request?

> I am not sure that we really need it.

If all periodic are canceled, users might need to re-create new ones.

> If a user is allowed to create an arbitrary number of timer events not linked to any request that may be a cause of DOS or CPU hog.

Maybe a directive like `js_periodic_max_events` is good.

> The alternative solution can be a shared queue that requests are able to push their task messages to be processed and and a single periodic handler per worker that processes the events.

It looks good.

Based on the above, we'll have two solutions.
1. 1 periodic: 1 message handler. the max number of periodicals is required.
2. 1 periodic: n message handlers. but it seems the number of max message handlers is also required.

Furthermore:
1. We could introduce a generic object `ngx.timer`.
2. Since it works in the worker process, so it can be independent of http context, we could introduce the `js_init_worker` directive. If it makes sense, I guess we need to change the implementation so that the fake `connection` and `request` is unnecessary. Anyway, it's a big change.
My thought is because of the feature, njs will be able to start running JS code in the worker process, the JS code contains the function `periodic`. And uses can use `ngx.fetch` together with `ngx.shared` to load something when worker processes start.

**Comment by xeioex** [comment]
@hongzhidao 

Thank you for the feedback.

>Since it works in the worker process, so it can be independent of http context, we could introduce the js_init_worker directive. If it makes sense, I guess we need to change the implementation so that the fake connection and request is unnecessary

For ngx.fetch() to have access to js_fetch_* directives it has be declared in a location, just like a [health_check](http://nginx.org/en/docs/http/ngx_http_upstream_hc_module.html#health_check) directive. But `js_init_worker` sounds like a good feature to have.

>We could introduce a generic object ngx.timer.

My issues with a dynamic timers (that is what we discussed with @ivanitskiy):

* it hard to understand and reason about the code that is half-in-nginx-conf and half-in-js-code. Ideally we need a single source of truth, and for nginx the nginx.conf looks a better alternative to me. with explicit js_periodic directives a user knows what is going on.
* I am yet need to find a use-case where we have to have dynamic timers instead of a periodic one.
* cancelling a timer in a single work process cancels it only in that process.
* more complex and leak-prone code.

>I feel (1) is more concise and generic.

Yes, the dynamic timers are more flexible, but what it actually gives in terms of performance, reliability and predictability under load? Looking from a pragmatic, not abstract software development perspective where more flexibility is good.
It is also easier to abuse by a user.

>I guess we need to change the implementation so that the fake connection and request is unnecessary. 

Right now having the fake request and face connection allows us to share  a bunch of [meta  handlers](https://github.com/nginx/njs/blob/master/nginx/ngx_http_js_module.c#L789) to reuse the existing code. As a result the patch is under 150 lines of code.

**Comment by hongzhidao** [comment]
> For ngx.fetch() to have access to js_fetch_* directives it has be declared in a location

It makes sense, and `js_imports` also has different http contexts.

> I am yet need to find a use-case where we have to have dynamic timers instead of a periodic one.

The `js_perodic` is to start a timer internally, it's good that you will support a way to stop it. I was thinking about how to start a new one if all the timers are canceled, but users can just reconfig nginx to start them. So it's ok to postpone the `js_init_worker` and dynamic timer. They are not necessary for now. The current way looks good, will look into it.

**Comment by mwcng** [comment]
Hi @xeioex 

Considering the nature of the new feature that you're implementing and it's ability to run independently on every worker process, would it be an appropriate time to ask you to reconsider implementing something similar to what I asked about in #609 ? 

Anything that would allow us to store data into an object (like js_preload_object does) without having to JSON.parse pretty hefty json strings on every single request would truly be amazing. It would help a lot in reducing server load for those who require reading a lot of dynamic/fast changing data on each request and on top of that it would also allow some other optimizations which would result in additional performance and functionality improvements.

I understand it's not an easy task based on what you said last time, but maybe this new feature that you're working on would make implementing something like this more approachable?

**Comment by drsm** [comment]
@xeioex 

What happens if a periodic timer overruns itself by using `setTimeout` for example?
And also in this case: 
```js
    async function handler() {
      return new Promice(() => {}); //never resolves
    }
```

**Comment by mwcng** [comment]
Hi @drsm 

I'm familiar with the shared dictionary that was added recently, it's definitely an awesome addition, however I didn't have the opportunity to test it in detail just yet.

Unless I'm misunderstanding something it can only support string + number zones and I'm essentially asking for `type=object` zones which would be able to store an object in the same manner that javascript does. To me it doesn't really matter if this is tied to the shared dictionary zone or if it's done outside of it. Anything that would provide `js_preload_object` functionality + the ability to update data periodically would be a very welcome addition.

I suppose using the shared dictionary I could create a bunch of keys to accomodate every single element from my current object + the nested objects like level0key_level1key_etc_etc that trickle down all the way to the value I need, but that feels kind of nightmarish to me and something that would potentially be very annoying to manage, which is why I think having something that can provide us with an updatable object would be an amazing addition.

**Comment by xeioex** [comment]
Hi @mwcng,

Thank you for the feedback. Agree, after js_periodic, it is a good time to reconsider the js_preload_object feature updates.

Having some API to update a js_preload_object seems doable. What bothers me is with js_periodic we would be able to update just a single worker VM. 

It means that during the update window the incoming client requests, that are randomly distributed across the workers will see the old and new data in js_preload_object. And this can be unexpected for some users.

**Comment by xeioex** [comment]
@drsm

>What happens if a periodic timer overruns itself by using setTimeout for example?

Good question, that is why I am asking some feedback to catch cases that I did not think through. I will look into these cases.

**Comment by xeioex** [comment]
@drsm 

>What happens if a periodic timer overruns itself by using setTimeout for example?

The current behaviour: there is only one periodic handler active per worker for a given `js_periodic`. if the handler produced the setTimeout() event, it may be any length of time. The interval parameter says the interval between the periodic events.

For example:
   with interval=60s, the handler may produce the 120s timeout, and the next periodic will start not in 1 but in 1 + 2 minutes.

Let me know whether it is a least surprising behaviour or not.

**Comment by drsm** [comment]
@xeioex 

So, just my thoughts, for someone who implies a rate when specifying `interval=60s` it would be surprising. Calculation of an average of, for example, `btc/eth` over a number of remote data sources every minute.

The "safe" :) strategy, I think, is to kill a running VM and restart a handler.

And, it would be nice to keep the current behavior as an option.

**Comment by hongzhidao** [comment]
@xeioex 

> Do you think it's necessary to make it programmable to stop a periodic?
Yes, agree this makes sense. Especially given that most of the time we need work in just one worker not in all, so we will be able just to disable all the periodics in other workers. I will add it.

It seems "reload" with nginx is a way to stop existing timers and start new ones. Maybe you don't need to support it, the current looks good.

**Comment by drsm** [comment]
@hongzhidao 
@xeioex 

Nginx on `reload` command performs a gracefull connection shutdown, should it also wait for currently running handlers?

**Comment by hongzhidao** [comment]
> Nginx on reload command performs a gracefull connection shutdown, should it also wait for currently running handlers?

It won't wait in the current implementation, the running worker processes stop. The periodic timer event is set to be cancelable.
```
+    periodic->event.cancelable = 1;
```

**Comment by drsm** [comment]
@hongzhidao 

Thanks.

So, it would be nice to have some hook to invoke when we're forcing exit from a running handler.
Something like:
```js
async function mayRunForawhile() {
    ngx.on('interrupt', (reason) => {
        // reason = "timeout" | "reload"
        ngx.log('periodic handler interrupted');
        // do cleanup, etc... 
    });
    // ...
}
```

**Comment by hongzhidao** [comment]
Btw, nginx has the `worker_shutdown_timeout` directive.
http://nginx.org/en/docs/ngx_core_module.html#worker_shutdown_timeout
It might be useful if we don't set the periodic event cancelable true.
The design of whether to set it cancelable is not clear to me yet, I'm thinking about it.

**Comment by drsm** [comment]
Probably, we should anyway cancel an execution, 'cause a broken code may run forever. And long running handlers is a normal situation. And there should be an easy way to not to leave garbage or things in an intermediate state.

**Comment by xeioex** [comment]
@drsm, @hongzhidao

> So, just my thoughts, for someone who implies a rate when specifying interval=60s it would be surprising. Calculation of an average of, for example, btc/eth over a number of remote data sources every minute (and data collection took more than minute due to dns lags).

Good point, thanks.

> The "safe" :) strategy, I think, is to kill a running VM and restart a handler.

I did it.

See the [updated](https://gist.github.com/xeioex/c66072ae30429bed693a0c7bd2660edf) patch:

1) it fires the js code at regular intervals, every "interval" seconds.

2) if the js periodic code takes too long to execute so it overlaps with the next interval, it is forcibly stopped and a new instance is started

3) I fixed the graceful shutdown on SIGQUIT. Previously if a JS code fires a very long timeout the nginx will never exit on SIGQUIT. Now  graceful shutdown is respected and after [worker_shutdown_timeout](http://nginx.org/en/docs/ngx_core_module.html#worker_shutdown_timeout) seconds, if a js periodic is still active it is forcibly finalized.

**Comment by hongzhidao** [comment]
Hi @xeioex 

No issues were found, and here are a few reworks. I'm also OK with the current way.
1. Use `ngx_array_t` for `jmcf->periodics` instead of it with a `next` field.
2. Get rid of `ngx_http_js_periodics_init()`.
3. It seems `ngx_terminate || ngx_exiting` is unnecessary, I checked other places that have a `cancelable` event.

The other question is about the stream module, is there any real use case for the periodic with stream module?

```
diff -r 7c8174ab92e3 nginx/ngx_http_js_module.c
--- a/nginx/ngx_http_js_module.c        Wed Aug 09 17:27:58 2023 -0700
+++ b/nginx/ngx_http_js_module.c        Mon Aug 21 15:26:33 2023 +0800
@@ -67,7 +67,6 @@
 struct ngx_js_periodic_s {
     ngx_http_conf_ctx_t   *conf_ctx;
     ngx_connection_t      *connection;
-    ngx_js_periodic_t     *next;

     /**
      * fd is used for event debug and should be at the same position
@@ -4103,10 +4102,6 @@

     periodic = ev->data;

-    if (ngx_terminate || ngx_exiting) {
-        return;
-    }
-
     timer = periodic->interval;

     if (periodic->jitter) {
@@ -4554,7 +4549,17 @@
         return NGX_OK;
     }

-    return ngx_http_js_periodics_init(jmcf->periodics);
+    periodic = jmcf->periodic->elts;
+
+    for (i = 0; i < = jmcf->periodic->nelts; i++) {
+        periodic->fd = 1000000 + i;
+
+        if (ngx_http_js_periodic_init(&periodic[i]) != NGX_OK) {
+            return NGX_ERROR;
+        }
+    }
+
+    return NGX_OK;
 }


diff -r 7c8174ab92e3 nginx/ngx_js.h
--- a/nginx/ngx_js.h    Wed Aug 09 17:27:58 2023 -0700
+++ b/nginx/ngx_js.h    Mon Aug 21 15:26:33 2023 +0800
@@ -46,7 +46,6 @@


 typedef struct ngx_js_dict_s  ngx_js_dict_t;
-typedef struct ngx_js_periodic_s  ngx_js_periodic_t;

 typedef struct {
     ngx_str_t              name;
@@ -58,7 +57,7 @@

 #define NGX_JS_COMMON_MAIN_CONF                                               \
     ngx_js_dict_t         *dicts;                                             \
-    ngx_js_periodic_t     *periodics                                          \
+    ngx_array_t           *periodics;                                         \


 #define _NGX_JS_COMMON_LOC_CONF                                               \
```

**Comment by hongzhidao** [comment]
Btw, is it good to merge the two patches into one?
```
# HG changeset patch
# User Dmitry Volyntsev <xeioex@nginx.com>
# Date 1691625711 25200
#      Wed Aug 09 17:01:51 2023 -0700
# Node ID 1f23b264507d5b6c8a8242ba5629a55140039b0f
# Parent  c4d0143a34a360cfd7c67b068c13ab766b19213e
Modules: introduced js_periodic directive.
```
```
# HG changeset patch
# User Dmitry Volyntsev <xeioex@nginx.com>
# Date 1691627278 25200
#      Wed Aug 09 17:27:58 2023 -0700
# Node ID e6512620410acbc44795912c434e75e63e5c4ef1
# Parent  1f23b264507d5b6c8a8242ba5629a55140039b0f
Modules: added support for gracesul shutdown in js_periodic.
```
There should be only one solution for the js periodical feature, right? It seems the third patch is an improvement of the second one, but they can be treated as the same thing.

**Comment by xeioex** [comment]
@hongzhidao 

Here is the latest: patch https://gist.github.com/xeioex/0d469fe881302816667f126458191853

It includes a separate stream directive and merges all the js_periodic directives together.

**Comment by xeioex** [comment]
@hongzhidao 

> It seems ngx_terminate || ngx_exiting is unnecessary, I checked other places that have a cancelable event.

BTW, it turned out that this is in fact necessary. Without it, the js_periodic handler would be called event after graceful shutdown signal and we may get in log the following message:

`2023/08/23 17:33:00 [alert] 15622#100160: *0 open socket #0 left in connection 4`

**Comment by hongzhidao** [comment]
@xeioex 
Good catch, you are right, it's necessary for the change.
> I fixed the graceful shutdown on SIGQUIT. Previously if a JS code fires a very long timeout the nginx will never exit on SIGQUIT. Now graceful shutdown is respected and after [worker_shutdown_timeout](http://nginx.org/en/docs/ngx_core_module.html#worker_shutdown_timeout) seconds, if a js periodic is still active it is forcibly finalized.
